### DISCUSSION TITLE
user_config_override.h // PCF8574

### DISCUSSION THREAD

**Discussion by nbranquinho** [arendst/Tasmota#17521] [Q&A] [answered]
I need help because something isn't ok, and I don't know what is.

I'm trying to put my PCF8574 working but I don't see the new menu in Tasmota web GUI ("Configure" => "Configure PCF8574").
I followed this [PCF8574 page](https://tasmota.github.io/docs/PCF8574/#introduction) but... I don't see anything.

I have

- Wemos D1 pro ( [like this one](https://www.google.com/url?sa=i&url=https%3A%2F%2Fsupport.envistiamall.com%2Fkb%2Fd1-mini-pro-nodemcu-and-arduino-wifi-lua-esp8266-wemos-microcontroller-with-antenna%2F&psig=AOvVaw07pLy3H7CzBCkgij_sPuG_&ust=1672172610855000&source=images&cd=vfe&ved=0CA8QjRxqFwoTCICOmsWOmPwCFQAAAAAdAAAAABAK) )
- 5v power supply 2A (to bredboard)
- Tasmota custom compilation ([followed this guide](https://tasmota.github.io/docs/Compile-your-build/) )
- PCF8574 connected to wemos
```
Wemos       PCF8574
  5V   <->  VCC
  G    <->  GND
  D1   <->  SCL
  D2   <->  SCA
```

The configuration is this:
![image](https://user-images.githubusercontent.com/2866871/209585458-d8c97c19-4f89-4dcc-a695-11460692f4b1.png)

In my _user_config_override.h_ file, I have this:
```

// !!! Remember that your changes GOES AT THE BOTTOM OF THIS FILE right before the last #endif !!!
*/

#ifdef CODE_IMAGE_STR
  #undef CODE_IMAGE_STR
#endif
#define CODE_IMAGE_STR "Test-esp8266-I2C-PCF8574"


#define USE_I2C                                // Add support for I2C
#define USE_PCF8574                            // [I2cDriver2] Enable PCF8574 I/O Expander (I2C addresses 0x20 - 0x26 and 0x39 - 0x3F) (+1k9 code)
#define USE_PCF8574_SENSOR                   // enable PCF8574 inputs and outputs in SENSOR message
#define USE_PCF8574_DISPLAYINPUT             // enable PCF8574 inputs display in Web page
#define USE_PCF8574_MQTTINPUT                // enable MQTT message & rule process on input change detection : stat/%topic%/PCF8574_INP = {"Time":"2021-03-07T16:19:23+01:00","PCF8574-1_INP":{"D1":1}}

```

After I do the command: I2Cdriver2 1, it restart but... nothing happen.

Result of STATUS 0:
`
15:39:43.167 RSL: STATUS = {"Status":{"Module":1,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_B2EF78","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
15:39:43.195 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"External System","Uptime":"0T00:00:11","StartupUTC":"2022-12-27T14:39:32","Sleep":50,"CfgHolder":4617,"BootCount":14,"BCResetTime":"2022-12-27T15:38:08","SaveCount":20,"SaveAddress":"F8000"}}
15:39:43.227 RSL: STATUS2 = {"StatusFWR":{"Version":"12.2.0(Test-esp8266-I2C-PCF8574)","BuildDateTime":"2022-12-27T11:59:10","Boot":23,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"354/699"}}
15:39:43.248 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["IotWifi",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
15:39:43.279 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":629,"Free":372,"Heap":27,"ProgramFlashSize":1024,"FlashSize":16384,"FlashChipId":"18401C","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000FB81","00004004","00001000","54000020","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,28,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
15:39:43.310 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-B2EF78-3960","IPAddress":"192.168.3.60","Gateway":"192.168.2.1","Subnetmask":"255.255.254.0","DNSServer1":"192.168.1.1","DNSServer2":"1.1.1.1","Mac":"A4:CF:12:B2:EF:78","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
15:39:43.335 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_B2EF78","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
15:39:43.356 RSL: STATUS7 = {"StatusTIM":{"UTC":"2022-12-27T14:39:43","Local":"2022-12-27T15:39:43","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"08:43","Sunset":"16:58"}}
15:39:43.375 RSL: STATUS10 = {"StatusSNS":{"Time":"2022-12-27T15:39:43"}}
15:39:43.382 RSL: STATUS11 = {"StatusSTS":{"Time":"2022-12-27T15:39:43","Uptime":"0T00:00:11","UptimeSec":11,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":25,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"IotWifi","BSSId":"F6:92:BF:98:ED:DD","Channel":13,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:04"}}}
`

Can someone help me?

Thanks a lot!!

**Comment by sfromis** [comment]
Well, I can see that you do have the PCF8574 support included in your build, the question may be if the driver is able to detect the device. You should see detection messages at boot time, and you can use the command `I2CScan` to check what I2C addresses are being found on the bus, should be in the intervals of `0x20-0x26` / `0x39-0x3F` depending on your address selection. If not, do check wiring carefully.

**Comment by nbranquinho** [comment, marked as answer]
I check the wires twice and is ok but when I run I2CScan, I get {"I2CScan":"No devices found"}.

`CMD: I2Cdriver
RSL: RESULT = {"I2CDriver":"2,7"}`

EDIT:

After a test with Arduino to check if HW is working well, I changed the address PINs (the 3 ones in the PCF8574 board) and voilá... start reporting... 
Now, tasmota see it :

` {"I2CScan":"Device(s) found at 0x3c"}`

So, it solved!

thank you @sfromis

**Reply by sfromis** [reply]
If your old address selection was `0x27`, that was (IIRC deliberately) excluded from the range as scanned by Tasmota.

**Reply by nbranquinho** [reply]
> If your old address selection was `0x27`, that was (IIRC deliberately) excluded from the range as scanned by Tasmota.

Checking the image, yes!! it point to 0x27 address and I don't remember that!!!! 
![image](https://user-images.githubusercontent.com/2866871/209943895-fa8d9496-218d-4476-9302-017c36750ec8.png)
( image taken from https://www.instructables.com/PCF8574-GPIO-Extender-With-Arduino-and-NodeMCU/ )

once again, thanks a lot 👍

**Reply by barbudor** [reply]
Addresses 0x27 and 0x3f are explicitly excluded and reserved for DISPLAY_LCD which are lcd displays based on PCF8574 but using another driver
Address 0x38 is excluded and reserved for other sensors

I'll soon submit a PR which will allow to easily override this defaults in custom builds

**Reply by nbranquinho** [reply]
> Addresses 0x27 and 0x3f are explicitly excluded and reserved for DISPLAY_LCD which are lcd displays based on PCF8574 but using another driver Address 0x38 is excluded and reserved for other sensors
> 
> I'll soon submit a PR which will allow to easily override this defaults in custom builds

Thx for the information!!!

**Reply by barbudor** [reply]
PR https://github.com/arendst/Tasmota/pull/17539 placed
Once accepted and merged you can override default search ranges to adapt the search range should you have a board where you cannot change the chip address.
